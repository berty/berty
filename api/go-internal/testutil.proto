syntax = "proto3";

package testutil;

option go_package = "berty.tech/berty/go/internal/testutil";

service TestService {
  // use to test unary
  rpc EchoTest(EchoTest.Request) returns (EchoTest.Reply);

  // Use to test stream.
  rpc EchoStreamTest(EchoStreamTest.Request) returns (stream EchoStreamTest.Reply);

  // Use to test duplex stream.
  rpc EchoDuplexTest(stream EchoDuplexTest.Request) returns (stream EchoDuplexTest.Reply);
}

message EchoTest {
  message Request {
    int64 delay = 1;
    string echo = 2;
    bool trigger_error = 3;
  }
  message Reply {
    string echo = 1;
  }
}

message EchoStreamTest {
  message Request {
    int64 delay = 1;
    string echo = 2;
    bool trigger_error = 3;
  }
  message Reply {
    string echo = 1;
  }
}

message EchoDuplexTest {
  message Request {
    string echo = 1;
    bool trigger_error = 2;
  }
  message Reply {
    string echo = 1;
  }
}
