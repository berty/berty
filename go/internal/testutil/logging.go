package testutil

import (
	"flag"
	"os"
	"strings"
	"sync"
	"testing"

	"github.com/stretchr/testify/assert"
	"go.uber.org/zap"

	"berty.tech/berty/v2/go/internal/logutil"
)

const defaultLoggingFilters = "info+:bty.test* error+:*,-ipfs*"

var (
	logFilters     = flag.String("log-filters", defaultLoggingFilters, "log namespaces")
	logFile        = flag.String("log-file", "", "log to file")
	logFormat      = flag.String("log-format", "color", "json, console, color")
	loggerInstance *zap.Logger
	loggerCleanup  func()
	loggerInitOnce sync.Once
)

func Logger(t *testing.T) (*zap.Logger, func()) {
	t.Helper()

	loggerInitOnce.Do(func() {
		if val := os.Getenv("BERTY_LOGFILTERS"); val != "" {
			*logFilters = val
		}
		if val := os.Getenv("BERTY_LOGFILE"); val != "" {
			*logFile = val
		}
		if val := os.Getenv("BERTY_LOGFORMAT"); val != "" {
			*logFormat = val
		}
		*logFilters = strings.ReplaceAll(*logFilters, ":default:", defaultLoggingFilters)

		var err error
		loggerInstance, loggerCleanup, err = logutil.NewLogger(*logFilters, *logFormat, *logFile)
		if !assert.NoError(t, err) {
			loggerInstance = zap.NewNop()
			loggerCleanup = func() {}
		}
	})

	return loggerInstance, loggerCleanup
}
