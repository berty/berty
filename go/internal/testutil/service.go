package testutil

import (
	context "context"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"go.uber.org/zap"
	"google.golang.org/grpc"

	"berty.tech/berty/v2/go/pkg/errcode"
	"berty.tech/weshnet/v2/pkg/grpcutil"
)

type service struct {
	logger *zap.Logger

	UnimplementedTestServiceServer
}

type Options struct {
	Logger *zap.Logger
}

func (o *Options) applyDefault() {
	if o.Logger == nil {
		o.Logger = zap.NewNop()
	}
}

func TestingNewServiceClient(ctx context.Context, t *testing.T, opts *Options) (*grpc.ClientConn, *grpc.Server) {
	t.Helper()

	srv := grpc.NewServer()
	svc := NewService(opts)
	RegisterTestServiceServer(srv, svc)

	l := grpcutil.NewBufListener(2048)

	cc, err := l.NewClientConn(ctx)
	assert.NoError(t, err)

	go func() {
		_ = srv.Serve(l.Listener)
	}()

	t.Cleanup(func() { l.Close() })

	return cc, srv
}

func NewService(opts *Options) TestServiceServer {
	opts.applyDefault()
	return &service{
		logger: opts.Logger,
	}
}

func (svc *service) EchoTest(_ context.Context, req *EchoTest_Request) (*EchoTest_Reply, error) {
	if req.TriggerError {
		return nil, errcode.ErrCode_ErrTestEcho
	}

	if req.Delay > 0 {
		time.Sleep(time.Duration(req.Delay) * time.Millisecond)
	}

	return &EchoTest_Reply{Echo: req.Echo}, nil
}

func (svc *service) EchoStreamTest(req *EchoStreamTest_Request, srv TestService_EchoStreamTestServer) error {
	if req.TriggerError {
		return errcode.ErrCode_ErrTestEcho
	}

	for {
		err := srv.Send(&EchoStreamTest_Reply{Echo: req.Echo})
		if err != nil {
			return errcode.ErrCode_ErrTestEchoSend.Wrap(err)
		}

		time.Sleep(time.Duration(req.Delay) * time.Millisecond)
	}
}

func (svc *service) EchoDuplexTest(srv TestService_EchoDuplexTestServer) error {
	for {
		req, err := srv.Recv()
		if err != nil {
			return errcode.ErrCode_ErrTestEchoRecv.Wrap(err)
		}

		if req.TriggerError {
			return errcode.ErrCode_ErrTestEcho
		}

		err = srv.Send(&EchoDuplexTest_Reply{
			Echo: req.Echo,
		})
		if err != nil {
			return errcode.ErrCode_ErrTestEchoSend.Wrap(err)
		}
	}
}
