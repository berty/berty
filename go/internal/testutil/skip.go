package testutil

import (
	"os"
	"testing"
)

func SkipSlow(t *testing.T) {
	t.Helper()
	if os.Getenv("SKIP_SLOW_TESTS") == "1" {
		t.Skip("slow test skipped")
	}
}
