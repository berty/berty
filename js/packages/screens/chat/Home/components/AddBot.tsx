import { BlurView } from '@react-native-community/blur'
import { Text, Icon } from '@ui-kitten/components'
import { Buffer } from 'buffer'
import React, { useEffect } from 'react'
import { View, TouchableOpacity, StyleSheet } from 'react-native'
import { useSelector } from 'react-redux'

import Avatar from '@berty/assets/logo/buck_berty_icon_card.svg'
import {
	HorizontalButtons,
	SecondaryButtonIconLeft,
	TertiaryButtonIconLeft,
} from '@berty/components'
import { UnifiedText } from '@berty/components/shared-components/UnifiedText'
import { useAppDimensions } from '@berty/contexts/app-dimensions.context'
import { useStyles } from '@berty/contexts/styles'
import { useAppDispatch } from '@berty/hooks'
import {
	PersistentOptionsKeys,
	selectPersistentOptions,
	setPersistentOption,
} from '@berty/redux/reducers/persistentOptions.reducer'
import { useThemeColor } from '@berty/store'
import messengerMethodsHooks from '@berty/store/methods'
import { base64ToURLBase64 } from '@berty/utils/convert/base64'

const AddBotBody: React.FC<{
	displayName: string
	link: string
	closeModal: () => void
}> = ({ displayName, link, closeModal }) => {
	const { row, text, margin, padding, border } = useStyles()
	const { scaleHeight } = useAppDimensions()
	const colors = useThemeColor()
	const persistentOptions = useSelector(selectPersistentOptions)
	const dispatch = useAppDispatch()
	const { call: requestContact, done, error } = messengerMethodsHooks.useContactRequest()
	const {
		reply: pdlReply,
		error: pdlError,
		call,
		done: pdlDone,
	} = messengerMethodsHooks.useParseDeepLink()

	useEffect(() => {
		call({ link })
	}, [call, link])

	useEffect(() => {
		if (done && !error) {
			closeModal()
		}
	}, [done, error, closeModal])

	const onClose = () => {
		dispatch(
			setPersistentOption({
				type: PersistentOptionsKeys.Suggestions,
				payload: {
					...persistentOptions.suggestions,
					[displayName]: {
						...persistentOptions.suggestions[displayName],
						state: 'skipped',
					},
				},
			}),
		)
		closeModal()
	}

	const onAdd = () => {
		if (pdlDone && !pdlError && pdlReply && pdlReply.link?.bertyId?.accountPk) {
			dispatch(
				setPersistentOption({
					type: PersistentOptionsKeys.Suggestions,
					payload: {
						...persistentOptions.suggestions,
						[displayName]: {
							...persistentOptions.suggestions[displayName],
							state: 'added',
							pk: base64ToURLBase64(
								Buffer.from(pdlReply.link.bertyId.accountPk).toString('base64'),
							),
						},
					},
				}),
			)
			requestContact({
				link,
			})
		}
	}

	return pdlReply?.link?.bertyId?.accountPk ? (
		<View
			style={[
				{
					justifyContent: 'center',
					alignItems: 'center',
					height: 250 * scaleHeight,
					top: '25%',
				},
				margin.big,
			]}
		>
			<View
				style={[
					{
						width: 110 * scaleHeight,
						height: 110 * scaleHeight,
						backgroundColor: colors['main-background'],
						justifyContent: 'center',
						alignItems: 'center',
						position: 'relative',
						top: 50 * scaleHeight,
						zIndex: 1,
						elevation: 7,
						shadowOpacity: 0.1,
						shadowRadius: 5,
						shadowColor: colors.shadow,
						shadowOffset: { width: 0, height: 3 },
					},
					border.radius.scale(60),
				]}
			>
				<Avatar width={125 * scaleHeight} height={125 * scaleHeight} />
			</View>
			<View
				style={[
					padding.horizontal.medium,
					padding.bottom.large,
					border.radius.large,
					border.shadow.huge,
					{ backgroundColor: colors['main-background'], shadowColor: colors.shadow },
				]}
			>
				<View style={[margin.top.scale(70 * scaleHeight)]}>
					<Icon
						name='info-outline'
						fill={colors['background-header']}
						width={60 * scaleHeight}
						height={60 * scaleHeight}
						style={[row.item.justify, padding.top.large]}
					/>
					<UnifiedText style={[text.align.center, padding.top.small, text.size.large, text.bold]}>
						{`ðŸ‘‹ ADD ${displayName}?`}
					</UnifiedText>
					<Text style={[text.align.center, padding.top.scale(20), padding.horizontal.medium]}>
						<UnifiedText style={[text.light]}>
							You don't have any contacts yet would you like to add the
						</UnifiedText>
						<UnifiedText style={[text.bold]}>{` ${displayName} `}</UnifiedText>
						<UnifiedText style={[text.light]}>to discover and test conversations?</UnifiedText>
					</Text>
				</View>
				<View style={[margin.top.medium, margin.horizontal.medium]}>
					<HorizontalButtons>
						<TertiaryButtonIconLeft name='close' onPress={onClose}>
							SKIP
						</TertiaryButtonIconLeft>
						<SecondaryButtonIconLeft onPress={onAdd}>ADD !</SecondaryButtonIconLeft>
					</HorizontalButtons>
				</View>
			</View>
		</View>
	) : null
}

export const AddBot: React.FC<{ displayName: string; link: string; closeModal: () => void }> = ({
	link,
	displayName,
	closeModal,
}) => {
	const { windowHeight } = useAppDimensions()

	return (
		<View style={[StyleSheet.absoluteFill, { elevation: 6, zIndex: 6 }]}>
			<TouchableOpacity
				activeOpacity={0}
				style={[
					StyleSheet.absoluteFill,
					{
						position: 'absolute',
						top: 0,
						bottom: 0,
						left: 0,
						right: 0,
						height: windowHeight,
					},
				]}
				onPress={() => closeModal()}
			>
				<BlurView style={[StyleSheet.absoluteFill]} blurType='light' />
			</TouchableOpacity>
			<AddBotBody displayName={displayName} link={link} closeModal={closeModal} />
		</View>
	)
}
