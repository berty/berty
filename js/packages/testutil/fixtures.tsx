import beapi from '@berty-tech/api'

const empty: beapi.messenger.IMedia = {
	metadataBytes: beapi.messenger.MediaMetadata.encode({}).finish(),
}

const media = {
	empty,
}

export default {
	media,
}
