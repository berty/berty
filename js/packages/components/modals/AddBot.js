import React, { useEffect } from 'react'
import {
	View,
	TouchableOpacity,
	TouchableWithoutFeedback,
	Text as TextNative,
	StyleSheet,
} from 'react-native'
import { Text, Icon } from '@ui-kitten/components'
import { useStyles } from '@berty-tech/styles'
import messengerMethodsHooks from '@berty-tech/store/methods'
import { useMsgrContext } from '@berty-tech/store/hooks'
import { useNavigation } from '@react-navigation/native'

import Avatar from './Buck_Berty_Icon_Card.svg'
import BlurView from '../common/BlurView'
import { PersistentOptionsKeys } from '@berty-tech/store/context'
import { base64ToURLBase64 } from '@berty-tech/components/utils'
import { Buffer } from 'buffer'

const useStylesAddBetabot = () => {
	const [{ width, border, padding, margin }] = useStyles()
	return {
		skipButton: [
			border.color.light.grey,
			border.scale(2),
			border.radius.small,
			margin.top.scale(15),
			padding.left.small,
			padding.right.medium,
			padding.top.small,
			padding.bottom.small,
			width(120),
		],
		addButton: [
			border.color.light.blue,
			border.scale(2),
			border.radius.small,
			margin.top.scale(15),
			padding.left.small,
			padding.right.medium,
			padding.top.small,
			padding.bottom.small,
			width(120),
		],
	}
}

export const AddBotBody = ({ displayName, link, closeModal }) => {
	const [
		{ row, text, margin, color, padding, background, border, opacity },
		{ scaleHeight },
	] = useStyles()
	const _styles = useStylesAddBetabot()
	const { setPersistentOption, persistentOptions } = useMsgrContext()
	const { call: requestContact, done, error } = messengerMethodsHooks.useContactRequest()
	const {
		reply: pdlReply,
		error: pdlError,
		call,
		done: pdlDone,
	} = messengerMethodsHooks.useParseDeepLink()

	useEffect(() => {
		call({ link })
	}, [call, link])

	useEffect(() => {
		if (done && !error) {
			closeModal()
		}
	}, [done, error, closeModal])

	return pdlReply?.link?.bertyId?.accountPk ? (
		<View
			style={[
				{
					justifyContent: 'center',
					alignItems: 'center',
					height: 250 * scaleHeight,
					top: 140,
				},
				margin.big,
			]}
		>
			<View
				style={[
					{
						width: 110 * scaleHeight,
						height: 110 * scaleHeight,
						backgroundColor: 'white',
						justifyContent: 'center',
						alignItems: 'center',
						position: 'relative',
						top: 50,
						zIndex: 1,
						shadowOpacity: 0.1,
						shadowRadius: 5,
						shadowOffset: { width: 0, height: 3 },
					},
					background.white,
					border.radius.scale(60),
				]}
			>
				<Avatar width={125 * scaleHeight} height={125 * scaleHeight} />
			</View>
			<View
				style={[
					background.white,
					padding.horizontal.medium,
					padding.bottom.medium,
					border.radius.large,
					border.shadow.huge,
				]}
			>
				<View style={[margin.top.scale(60)]}>
					<Icon
						name='info-outline'
						fill={color.blue}
						width={60 * scaleHeight}
						height={60 * scaleHeight}
						style={[row.item.justify, padding.top.large]}
					/>
					<TextNative
						style={[
							text.align.center,
							padding.top.small,
							text.size.large,
							text.bold.medium,
							text.color.black,
							{ fontFamily: 'Open Sans' },
						]}
					>
						{`ðŸ‘‹ ADD ${displayName}?`}
					</TextNative>
					<Text style={[text.align.center, padding.top.scale(20), padding.horizontal.medium]}>
						<TextNative
							style={[
								text.bold.small,
								text.size.medium,
								text.color.black,
								{ fontFamily: 'Open Sans' },
							]}
						>
							You don't have any contacts yet would you like to add the
						</TextNative>
						<TextNative
							style={[
								text.bold.medium,
								text.size.medium,
								text.color.black,
								{ fontFamily: 'Open Sans' },
							]}
						>
							{displayName}
						</TextNative>
						<TextNative
							style={[
								text.bold.small,
								text.color.black,
								text.size.medium,
								{ fontFamily: 'Open Sans' },
							]}
						>
							to discover and test conversations?
						</TextNative>
					</Text>
				</View>
				<View style={[row.center, padding.top.medium]}>
					<TouchableOpacity
						style={[row.fill, margin.bottom.medium, opacity(0.5), _styles.skipButton]}
						onPress={async () => {
							await setPersistentOption({
								type: PersistentOptionsKeys.Suggestions,
								payload: {
									...persistentOptions.suggestions,
									[displayName]: {
										...persistentOptions.suggestions[displayName],
										state: 'skipped',
									},
								},
							})
							closeModal()
						}}
					>
						<Icon name='close' width={30} height={30} fill={color.grey} style={row.item.justify} />
						<TextNative
							style={[
								text.color.grey,
								padding.left.small,
								row.item.justify,
								text.size.scale(16),
								text.bold.medium,
								{ fontFamily: 'Open Sans' },
							]}
						>
							SKIP
						</TextNative>
					</TouchableOpacity>
					<TouchableOpacity
						style={[row.fill, margin.bottom.medium, background.light.blue, _styles.addButton]}
						onPress={async () => {
							if (pdlDone && !pdlError) {
								console.log('heeere')
								await setPersistentOption({
									type: PersistentOptionsKeys.Suggestions,
									payload: {
										...persistentOptions.suggestions,
										[displayName]: {
											...persistentOptions.suggestions[displayName],
											state: 'added',
											pk: base64ToURLBase64(
												new Buffer(pdlReply.link.bertyId.accountPk).toString('base64'),
											),
										},
									},
								})
								await requestContact({
									link,
								})
							}
						}}
					>
						<Icon
							name='checkmark-outline'
							width={30}
							height={30}
							fill={color.blue}
							style={row.item.justify}
						/>
						<TextNative
							style={[
								text.color.blue,
								padding.left.small,
								row.item.justify,
								text.size.scale(16),
								text.bold.medium,
							]}
						>
							ADD !
						</TextNative>
					</TouchableOpacity>
				</View>
			</View>
		</View>
	) : null
}

export const AddBot = ({ link, displayName, closeModal }) => {
	const [{}, { windowHeight }] = useStyles()

	return (
		<View style={[StyleSheet.absoluteFill]}>
			<TouchableOpacity
				activeOpacity={0}
				style={[
					StyleSheet.absoluteFill,
					{
						position: 'absolute',
						top: 0,
						bottom: 0,
						left: 0,
						right: 0,
						height: windowHeight,
					},
				]}
				onPress={() => closeModal()}
			>
				<BlurView style={[StyleSheet.absoluteFill]} blurType='light' />
			</TouchableOpacity>
			<AddBotBody displayName={displayName} link={link} closeModal={closeModal} />
		</View>
	)
}

export default AddBot
