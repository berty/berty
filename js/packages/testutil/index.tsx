import React from 'react'
import { NavigationContainer } from '@react-navigation/native'
import { IconRegistry } from '@ui-kitten/components'
import { EvaIconsPack } from '@ui-kitten/eva-icons'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import { Provider as ThemeProvider } from '@berty-tech/components/theme'
import { FeatherIconsPack } from '@berty-tech/messenger-app/feather-icons'
import { CustomIconsPack } from '@berty-tech/messenger-app/custom-icons'
import { Provider as StyleProvider } from '@berty-tech/styles'

export const TestProvider: React.FC = ({ children }) => (
	<SafeAreaProvider
		initialMetrics={{
			frame: { x: 0, y: 0, width: 0, height: 0 },
			insets: { top: 0, left: 0, right: 0, bottom: 0 },
		}}
	>
		<StyleProvider>
			<IconRegistry icons={[EvaIconsPack, FeatherIconsPack, CustomIconsPack]} />
			<ThemeProvider>
				<NavigationContainer>{children}</NavigationContainer>
			</ThemeProvider>
		</StyleProvider>
	</SafeAreaProvider>
)

import _fixtures from './fixtures'
export const fixtures = _fixtures
