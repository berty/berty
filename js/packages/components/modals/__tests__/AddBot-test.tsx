import React from 'react'
import renderer from 'react-test-renderer'

import { TestProvider } from '@berty-tech/testutil'

import AddBot from '../AddBot'

test('AddBot modal renders correctly', () => {
	const props: React.ComponentProps<typeof AddBot> = {
		displayName: 'Alice',
		link: 'https://berty.tech/id#contact/oZBLFWJdsE73yWa2BjceiUtKCfxpxVBuH8h6r3sJXDrgqaHyR1ACgLpWEZTMDD8DzGsSo679UjknhcuXGiqP6cgU66R5BTh/name=Alice',
		closeModal: () => {},
	}
	// FIXME: mock useContactRequest to get children in the snapshot
	const tree = renderer
		.create(
			<TestProvider>
				<AddBot {...props} />
			</TestProvider>,
		)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
