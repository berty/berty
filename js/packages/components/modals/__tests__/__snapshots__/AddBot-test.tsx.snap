// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBot modal renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "elevation": 6,
            "zIndex": 6,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "bottom": 0,
            "height": 1334,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <BlurView
          blurAmount={10}
          blurType="light"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ],
            ]
          }
        />
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
